alloc.h

```diff
@@ -6,4 +6,19 @@
 #define alloc(x) malloc(x)
 #define alloc_free(x) free(x)
 
+/*
+ * This functions should not be used in any new code. It is provided so that
+ * external patches that rely on it's presence still work correctly.
+ */
+#ifndef DONT_PROVIDE_OBSOLETE
```

Comment:
Double-negative conditions are hard for my brain. What are our goals for people and code trying to interact with this function? I’m thinking once we ship 1.09 with intent-to-deprecate, we want

1. Zero callers in our tree
2. Not exposed by default, so we don’t start calling it by mistake
3. If someone has a patch that needs `alloc_re()`, they’ll have to `#define` something to get it

Before 1.09 ships, seems like it suffices to preserve the API as it was (and add a big warning comment like you’ve done) — unless you’re hoping to start accomplishing (2) now?

Reply:
The only place I can see where DJB uses alloc_re() outside of gen_allocdefs.h in any of his projects is in env.c where it's used to add data into char **environ. Not only should manually calling alloc_re() be used sparingly, this patch and env.c may benefit from having their memory management logic revised. GEN_ALLOC_typedef() is used to provision more than just stralloc data structures, so I think that envalloc and resolvealloc typedefs may be suitable here just to standardize how notqmail handles dynamically allocating memory, unless someone thinks strallocs are a bad idea.

Reply:
These places are already changed to use plain ```realloc()```.

Reply:
The intention here is indeed that we can build (and one CI run now does build) without that function being available, to enforce that we have no users in tree.

Reply:
Glad to hear it. The question then is, apply the patch as is, apply the patch with realloc(), or standardize allocations globally and apply the patch with the standardized API?

Reply:
If we are going to apply the large DNS patch (I think you mean that with "this patch"), then we will for sure fix it to use realloc() instead. This commit is needed for everything that may else be added by patch by users and which suddenly may fail to work otherwise.

Reply:
My preferred course, I think, is to start with this:

```c
#if defined(I_KNOW_THIS_DEPRECATED_CALL_WILL_GO_AWAY_SOON_AND_I_PROMISE_TO_FIX_MY_CALLING_CODE) \
    || !defined(REMOVE_DEPRECATED_CODE)
...
#endif 
```

(Ha-ha only serious about the first macro really being called that. That’s pretty much exactly what I did last time I had to migrate API callers, and it really did help people understand the implications of what they were doing.)

In the same PR branch, have TravisCI define `REMOVE_DEPRECATED_CODE` so we can start seeing how that goes.

Once 1.09 is released, remove the `|| !defined(REMOVE_DEPRECATED_CODE)` (and the now-superfluous TravisCI setting).

Reply:
I just marked them formally as deprecated using compiler attributes and turned the usage of those functions to compiler errors in the CI. How about that?

Reply:
Yeah, that’s probably best. Does it give the build-time behaviors we want, if so, great.